Hao Wei - Update DG and PPP (#206)
AY2122S2-CS2103-W17-1/tp: docs/DeveloperGuide.md

This section describes some noteworthy details on how certain features are implemented.

### Tag Management

This section will describe tag management in the address book as well as the features implemented.

### Centralising Tags in the Address Book

> In the previous implementation, all `Tag`s are independent of one another despite having the same tag names. Hence, centralising the tags enable the user to easily manipulate them as well as easily searching for `Person`s that contain the tag.

Reviewer: I think when adding the `Tag` or `Person`, try not to follow it up with a "s" behind cos when it is rendered it looks a bit weird.
<img width="65" alt="Screenshot 2022-03-21 at 10 42 16 PM" src="https://user-images.githubusercontent.com/13099969/159285397-a8171d95-61c0-4d6f-b2bc-fc86f7bf32dc.png">
 
I can suggest referring it to as `Tag` objects or `Person` objects

Reviewer: I realise this applies to many other places in this PR so I'm not gonna individually highlight each of them lol if you decide to alter this

Author: Thank you for your feedback on this! I have followed your suggestion and amended the changes in 61207721

> In the previous implementation, all `Tag`s are independent of one another despite having the same tag names. Hence, centralising the tags enable the user to easily manipulate them as well as easily searching for `Person`s that contain the tag.

Reviewer 2: > Hence, centralising ...

There is nothing to continue from, `hence` does not make sense in this context.

Author: Rephrased in 61207721

> This is done by creating a `UniqueTagList` within `AddressBook` which will store all tags that were created by the user. Whenever a command relating to `Tag` is executed, it will not only apply the changes to the `Tag` in the `UniqueTagList` but will propagate these changes to the relevant `Person`s who contain the specified tag.

Reviewer 2: I suggest a small english fix to make the text more fluent.
```suggestion
This is done by creating a `UniqueTagList` within `AddressBook` which will store all tags that were created by the user. Whenever a command relating to `Tag` is executed, it will not only apply the changes to the `Tag` in the `UniqueTagList` but will also propagate these changes to the relevant `Person`s who contain the specified tag.
```

> Hence, all operations relating to the `Tag`s are done at the `AddressBook` level to ensure that the `Tag`s and `Person`s are properly synchronised.

Reviewer 2: Once again, `hence` does not really make sense in this context, you should drop it.
```suggestion
All operations relating to the `Tag`s are done at the `AddressBook` level to ensure that the `Tag`s and `Person`s are properly synchronised.
```

Author: Fixed in 61207721

> Another benefit that comes with the centralised tag list is that the user can maintain tags even if it is not associated to any `Person`s. The rationale to maintain `Tag` is to allow the user to reuse the tag depending on their workflow (i.e. A user may want to maintain the `prospective clients` tag even if he/she currently does not have any prospective clients.)

Reviewer 2: Small grammatical issue here.
```suggestion
Another benefit that comes with the centralised tag list is that the user can maintain tags even if it is not associated with any `Person`s. The rationale to maintain `Tag` separately is to allow the user to reuse the tag depending on their workflow (i.e. A user may want to maintain the `prospective clients` tag even if he/she currently does not have any prospective clients.)
```

### Edit Tag Feature - `edittag`

The tag editing feature is similar to the system used for `Person` but extended to propagate the changes to the `Person`s. This feature is implemented at the `AddressBook` level, and the related functions are:

* `AddressBook#setTag(target, editedTag)`
* `AddressBook#setPersonsWithTag(target, editedTag)`

Note: `target` refers to the tag to be updated, and `editedTag` is the replacement tag specified by the user.

![Tag Edit](images/TagEditSequenceDiagram.png)

### Serialisation and Inflation

> `Tag` serialisation and inflation is handled by the `Storage` component. The current implementation extends the existing method from `JsonSerializableAddressBook` to read a list of tag names from the JSON file and saving them as its own JSON file.

Reviewer 2: I understand what you mean when you say extend, but technically extend should refer to sub-classing (which you are not actually doing). Maybe this phrasing is better?
```suggestion
`Tag` serialisation and inflation is handled by the `Storage` component. The current implementation augments the existing method from `JsonSerializableAddressBook` through the addition of reading a list of tag names from the JSON file and saving them as its own JSON file.
```

Author: Fixed in 61207721

#### Serialisation of Tags

Since the tags are independent to the `Person`, the serialisation does not require any special attention for the dependency, as the integrity is guaranteed by the `AddressBook` component.

#### Inflation of Tags

To ensure the `Tag`s are properly added into the address book, `JsonSerializableAddressBook#toModelType()` has been modified to inflate the tags first before the person. This is to ensure that duplicate tags are not added into the address book by accident and will only add tags that do not exist in the tag list (which could be caused by the user manually adding the tags in the user-editable JSON file).

> Hence, the `JSONSerializableAddressBook#addMissingTags()` is implemented to check all `Tag`s within each `Person` and add only the missing `Tag`s.

Reviewer 2: Once again, hence is weirdly used in this context.
```suggestion
A helper method `JSONSerializableAddressBook#addMissingTags()` is implemented to check all `Tag`s within each `Person` and add only the missing `Tag`s.
```
